jf_agent/main.py

```diff
@@ -366,37 +370,53 @@ def download_data(config, creds, endpoint_jira_info, endpoint_git_instances_info
             print_all_jira_fields(config, jira_connection)
         download_data_status.append(load_and_dump_jira(config, endpoint_jira_info, jira_connection))
 
-    is_multi_git_config = len(config.git_configs) > 1
-    for git_config in config.git_configs:
-        agent_logging.log_and_print(
-            logger,
-            logging.INFO,
-            f'Obtained {git_config.git_provider} configuration, attempting download...',
-        )
-        if is_multi_git_config:
-            instance_slug = git_config.git_instance_slug
-            instance_info = endpoint_git_instances_info.get(instance_slug)
-            instance_creds = creds.git_instance_to_creds.get(instance_slug)
-        else:
-            # support legacy single-git support, which assumes only one available git instance
-            instance_info = list(endpoint_git_instances_info.values())[0]
-            instance_creds = list(creds.git_instance_to_creds.values())[0]
+    
+    # git downloading is parallelized by the number of configurations. 
+    futures = []
+    with ThreadPoolExecutor(max_workers=8) as executor:
+        for git_config in config.git_configs:
+            agent_logging.log_and_print(
+                logger,
+                logging.INFO,
+                f'Obtained {git_config.git_provider}:{git_config.git_instance_slug} configuration, attempting download {"in parallel" if len(config.git_configs) > 1 else ""}...',
+            )
+            futures.append(
+                executor.submit(
+                    _download_git_data,
+                    git_config,
+                    config,
+                    creds,
+                    endpoint_git_instances_info,
+                    len(config.git_configs) > 1,
+                )
+            )
 
-        git_connection = get_git_client(
-            git_config, instance_creds, skip_ssl_verification=config.skip_ssl_verification
-        )
+    return download_data_status + [f.result() for f in futures]
```

Comment:
If you haven't already, I'd double-check that the exception handling is what you'd want it to be -- i.e., if an exception propagates out of one of the threads, it's raised in the main thread (rather than being swallowed).

Reply:
The `concurrent.Future.result()` method will raise the exception if one occurred on the thread. 
https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result
I tried this out by throwingn in `_download_git_data` and it works as expected.